code/datums/repositories/decls.dm

```diff
@@ -17,26 +17,39 @@
 var/global/repository/decls/decls_repository = new
 
 /repository/decls
-	var/list/fetched_decls
-	var/list/fetched_decl_types
-	var/list/fetched_decl_subtypes
+	var/list/fetched_decls =         list()
+	var/list/fetched_decl_ids =      list()
+	var/list/fetched_decl_types =    list()
+	var/list/fetched_decl_subtypes = list()
 
 /repository/decls/New()
 	..()
-	fetched_decls = list()
-	fetched_decl_types = list()
-	fetched_decl_subtypes = list()
+	for(var/decl_type in typesof(/decl))
+		var/decl/decl = decl_type
+		var/decl_uid = initial(decl.uid)
+		if(decl_uid)
+			fetched_decl_ids[decl_uid] = decl_type
+
+/repository/decls/proc/get_decl_by_id(var/decl_id)
+	. = get_decl(fetched_decl_ids[decl_id])
 
 /repository/decls/proc/get_decl(var/decl_type)
 	ASSERT(ispath(decl_type))
 	. = fetched_decls[decl_type]
 	if(!.)
-		. = new decl_type()
-		fetched_decls[decl_type] = .
-
-		var/decl/decl = .
-		if(istype(decl))
-			decl.Initialize()
+		var/decl/decl = new decl_type()
+		if(decl_type == decl.abstract_type && decl.crash_on_abstract_init)
```

Comment:
Should this be using the `is_abstract` helper?

Reply:
Sure.